Choose mag when training models on multiple annotations

knollengewaechs opened the pull request:
### URL of deployed dev instance (used for testing):
- https://___.webknossos.xyz

### Steps to test:
- Enable jobs
- Open an annotation and go to AI Analysis > Train a model
- Select a ground truth and a color layer
- Select a mag. Check that all available mags are present in both layers
- Start a training / check network tab that the right mag was sent in the request
- Do the same thing, but for multiple annotations. Do that by navigating to Administration > AI models > Train new model.


- ### TODOs:
- [x] update mags after any of the input layers change 
    - [x] deselect mag
    - [x] move value to upper component
- [x] find union of data and ground truth layer
    - [x] datastructure: save map annotation id -> datalayermags, groundtruthlayermags
- [x] check that it works from within annotation and with multiple annotations
- [x] fix spacing (required vs not required fields in form, see pic)
- [x] address first review
- [ ] regarding testing: does this need to be tested after the request was sent? if yes, how?
- [x] when opening the modal, there is a ground truth layer option for the active segmentation layer, and for the fallback layer. the first one's name is simply the tracing id, which makes it harder to understand what is happening. 
![image](https://github.com/user-attachments/assets/774b507a-62fe-4aae-a3ed-3712a77b361c)



### Issues:
- contributes to #8097

------
(Please delete unneeded items, merge only when none are left open)
- [x] Updated [changelog](../blob/master/CHANGELOG.unreleased.md#unreleased)
- [ ] Updated [migration guide](../blob/master/MIGRATIONS.unreleased.md#unreleased) if applicable
- [ ] Updated [documentation](../blob/master/docs) if applicable
- [ ] Adapted [wk-libs python client](https://github.com/scalableminds/webknossos-libs/tree/master/webknossos/webknossos/client) if relevant API parts change
- [ ] Removed dev-only changes like prints and application.conf edits
- [x] Considered [common edge cases](../blob/master/.github/common_edge_cases.md)
- [ ] Needs datastore update after deployment

knollengewaechs commented on conf/application.conf:
```diff
@@ -151,8 +151,8 @@ features {
   # to enable jobs for local development, use "yarn enable-jobs" to also activate it in the database
-  jobsEnabled = false
-  voxelyticsEnabled = false
+  jobsEnabled = true
+  voxelyticsEnabled = true
```
remove me before merging!!!

knollengewaechs commented on frontend/javascripts/oxalis/view/jobs/train_ai_model.tsx:
```diff
@@ -311,19 +393,52 @@ export function TrainAiModelTab<GenericAnnotation extends APIAnnotation | Hybrid
                 getReadableNameForLayer={(layer) => {
+                  console.log(layer);
                   return layer.name;
+                  //TODO_c fix that fallback layers are shown at least with the correct name?
+                  // eg. name (fallbacklayer)
+                  // remember to check-in with P
```
also see PR description. to me this feels like a separate issue

knollengewaechs commented on frontend/javascripts/components/mag_selection.tsx:
```diff
@@ -0,0 +1,73 @@
+    <Select
+      placeholder="Select a magnification"
+      value={
+        // using the index of the mag *in the mag list* as value internally
+        // NB: this is different from the mag index
+        value == null || magInfo == null
+          ? null
+          : clamp(
+              0,
+              allMags.findIndex((v) => V3.equals(v, value)),
+              allMags.length - 1,
+            )
+      }
```
the value will be found because only present values can be selected

knollengewaechs commented on frontend/javascripts/oxalis/view/jobs/train_ai_model.tsx:
```diff
@@ -311,19 +393,52 @@ export function TrainAiModelTab<GenericAnnotation extends APIAnnotation | Hybrid
+                      form.getFieldValue(["trainingAnnotations", idx, "layerName"]),
+                      form.getFieldValue(["trainingAnnotations", idx, "imageDataLayer"]),
+                    );
+                    form.setFieldValue(["trainingAnnotations", idx, "mag"], undefined);
```
https://ant.design/components/form says `Use setFieldValue instead if you want to only config single value in Form.List` for `setFieldsValue` so I think its fine the way it is

philippotto commented on frontend/javascripts/oxalis/view/jobs/train_ai_model.tsx:
```diff
@@ -311,19 +388,50 @@ export function TrainAiModelTab<GenericAnnotation extends APIAnnotation | Hybrid
                 fixedLayerName={fixedSelectedSegmentationLayer?.name || undefined}
                 label="Ground Truth Layer"
+                onChange={() => {
+                  setIntersectingMags(
+                    annotationId,
+                    dataset,
+                    form.getFieldValue(["trainingAnnotations", idx, "layerName"]),
+                    form.getFieldValue(["trainingAnnotations", idx, "imageDataLayer"]),
+                  );
+                  form.setFieldValue(["trainingAnnotations", idx, "mag"], undefined);
```
seems to be a copy of the other onchange handler. I'd extract it into an "onChangeLayer" handler.

philippotto commented on frontend/javascripts/oxalis/view/jobs/train_ai_model.tsx:
```diff
@@ -150,21 +158,78 @@ export function TrainAiModelFromAnnotationTab({ onClose }: { onClose: () => void
+      groundTruthLayerName,
+      imageDataLayerName,
+    );
+    if (mags == null) {
+      return [{ annotationId, magInfo: new MagInfo(intersectingMags) }];
+    } else {
+      return mags.map((mag) =>
+        mag.annotationId === annotationId
+          ? { annotationId, magInfo: new MagInfo(intersectingMags) }
+          : mag,
+      );
+    }
```
tbh, I don't understand the reasoning behind this code.

philippotto reviewed (commented):
sorry, I started reviewing the PR a few days ago and then apparently something else grabbed my attention. I finished the review now and left a few comments.

It's cool that the mag selection seems to work :+1: however, I find the code a bit confusing. I think my main gripe with it is that `mags` can be null and that it is written only on certain events. In my mind, all information should be available at all times and mags could always be computed on demand (memoized if necessary). I understand that antd forms can be annoying, so maybe it doesn't work how I imagine it, but please let's have a call to discuss this :)

knollengewaechs commented:
I think I addressed all the feedback for now :crossed_fingers:  From my side an open question that is left is how to handle volume annotation layer names (=tracing ids from what I understand) in the selection. Maybe something like this? 
![image](https://github.com/user-attachments/assets/0b6b00bf-064d-4fcb-8512-1480766537e8)

philippotto commented on frontend/javascripts/oxalis/view/jobs/train_ai_model.tsx:
```diff
@@ -250,8 +251,8 @@ export function TrainAiModelTab<GenericAnnotation extends APIAnnotation | Hybrid
     }
 
-    await runTraining({
-      trainingAnnotations: await getTrainingAnnotations(values),
```
was this on purpose? you re-added the first await again [here](https://github.com/scalableminds/webknossos/pull/8266/commits/63258289b8889f5628300b93c5a3c93d1bd43a0f), but the second one not? I think, both awaits are important?
